src/browser/api/window.js

```diff
@@ -1868,56 +1852,90 @@ function emitReloadedEvent(identity, url) {
+            if (browserWindow._options.saveWindowState) {
+                const cachedBounds = _boundsChangedHandler.getCachedBounds();
+                saveBoundsToDisk(identity, cachedBounds, err => {
+                    if (err) {
+                        log.writeToLog('info', err);
+                    }
+                    // These were causing an exception on close if the window was reloaded
+                    _boundsChangedHandler.teardown();
+                    resolve();
+                });
+            } else {
+                _boundsChangedHandler.teardown();
+                resolve();
+            }
+        });
+    }
+
+    //Window tear down will:
+    //    Update core state by removing the window.
+    //    Save the window state to disk
+    //    Close all child windows
+    //    Wait for the close event.
     return function() {
         let ofWindow = Window.wrap(identity.uuid, identity.name);
-        let childWindows = coreState.getChildrenByWinId(id);
-
+        let childWindows = coreState.getChildrenByWinId(id) || [];
         // remove from core state earlier rather than later
         coreState.removeChildById(id);
 
         // remove window from any groups it belongs to
         WindowGroups.leaveGroup(ofWindow);
 
-        if (childWindows && childWindows.length > 0) {
-            let closedChildren = 0;
+        promises.push(handleSaveStateAlwaysResolve());
 
-            childWindows.forEach(childId => {
-                let child = coreState.getWinObjById(childId);
-
-                // TODO right now this is forceable to handle the event that there was a close
-                //      requested on a child window and the main window closes. This needs
-                //      looking into
-                if (child) {
-                    let childIdentity = {
-                        name: child.name,
-                        uuid: child.uuid
-                    };
+        childWindows.forEach(childId => {
+            promises.push(closeChildWin(childId));
+        });
 
-                    Window.close(childIdentity, true, () => {
-                        closedChildren++;
-                        if (closedChildren === childWindows.length) {
-                            emitCloseEvents(identity);
-                            coreState.removeChildById(id);
-                        }
-                    });
-                } else {
-                    closedChildren++;
-                    if (closedChildren === childWindows.length) {
-                        emitCloseEvents(identity);
-                        coreState.removeChildById(id);
-                    }
-                }
-            });
-        } else {
+        return Promise.all(promises).then(() => {
             emitCloseEvents(identity);
-        }
+            browserWindow.removeAllListeners();
+            if (childWindows.length > 0) {
+                coreState.removeChildById(id);
```

Comment:
Why is `coreState.removeChildById` being called here again? It's already called here https://github.com/HadoukenIO/core/pull/485/files?utf8=%E2%9C%93&diff=split#diff-d5752a5b2c3b7ce27c77d78f1b9278a6R1915

Reply:
If there's a reason it should stay, we should leave a comment explaining why.